src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs

```diff
@@ -0,0 +1,527 @@
+                    parentProjectSpec.RestoreMetadata?.ProjectWideWarningProperties,
+                    PackageSpecificWarningProperties.CreatePackageSpecificWarningProperties(parentProjectSpec),
+                    parentProjectSpec.TargetFrameworks.Select(f => f.FrameworkName).AsList().AsReadOnly());
+
+            foreach (var targetGraph in targetGraphs)
+            {
+                if (string.IsNullOrEmpty(targetGraph.RuntimeIdentifier))
+                {
+                    var transitiveNoWarnFromTargetGraph = ExtractTransitiveNoWarnProperties(
+                        targetGraph,
+                        parentProjectSpec.RestoreMetadata.ProjectName,
+                        parentWarningProperties);
+
+                    projectFrameworks.Add(targetGraph.Framework);
+
+                    transitivePackageSpecificProperties = MergePackageSpecificWarningProperties(
+                        transitivePackageSpecificProperties, 
+                        transitiveNoWarnFromTargetGraph);
+                }
+            }
+
+            return new WarningPropertiesCollection(
+                projectWideWarningProperties: null,
+                packageSpecificWarningProperties: transitivePackageSpecificProperties,
+                projectFrameworks: projectFrameworks
+                );
+        }
+
+        /// <summary>
+        /// Traverses a Dependency grpah starting from the parent project in BF style.
+        /// </summary>
+        /// <param name="targetGraph">Parent project restore target graph.</param>
+        /// <param name="parentProjectName">File path of the parent project.</param>
+        /// <param name="parentWarningPropertiesCollection">WarningPropertiesCollection of the parent project.</param>
+        /// <returns>PackageSpecificWarningProperties containing all the NoWarn's for each package seen in the graph accumulated while traversing the graph.</returns>
+        private PackageSpecificWarningProperties ExtractTransitiveNoWarnProperties(
+            RestoreTargetGraph targetGraph,
+            string parentProjectName,
+            WarningPropertiesCollection parentWarningPropertiesCollection)
+        {
+            var dependencyMapping = new Dictionary<string, LookUpNode>(StringComparer.OrdinalIgnoreCase);
+            var queue = new Queue<DependencyNode>();
+            var seen = new HashSet<DependencyNode>();
+            var frameworkReducer = new FrameworkReducer();
+            var resultWarningProperties = new PackageSpecificWarningProperties();
+            var packageNoWarn = new Dictionary<string, ISet<NuGetLogCode>>(StringComparer.OrdinalIgnoreCase);
+
+            // All the packages in parent project's closure. 
+            // Once we have collected data for all of these, we can exit.
+            var parentPackageDependencies = new HashSet<string>(
+                targetGraph.Flattened.Where(d => d.Key.Type == LibraryType.Package).Select(d => d.Key.Name));
```

Comment:
This should be created as part of the flattened loop below, it isn't used until later

Reply:
This also looks incorrect since it is case sensitive by default, if this only checks the exact id because there can only be 1 instance of the id then it should use Ordinal instead of the culture specific default.